File: packages/mp_flutter_runtime/lib/components/component_factory.dart
part of '../mp_flutter_runtime.dart';

class _MPComponentFactory {
  final MPEngine engine;
  final List<Map> _textMeasureResults = [];
  final Map<int, ComponentViewState> _cacheViews = {};

  _MPComponentFactory({required this.engine});

  Widget create(Map? data, {Map? parentData}) {
    if (data != null) {
      String? name = data['name'];
      if (name != null) {
        switch (name) {
          case 'absorb_pointer':
            return _AbsorbPointer(
                data: data, parentData: parentData, componentFactory: this);
          case 'clip_oval':
            return _ClipOval(
                data: data, parentData: parentData, componentFactory: this);
          case 'clip_r_rect':
            return _ClipRRect(
                data: data, parentData: parentData, componentFactory: this);
          case 'colored_box':
            return _ColoredBox(
                data: data, parentData: parentData, componentFactory: this);
          case 'custom_paint':
            return _CustomPaint(
                data: data, parentData: parentData, componentFactory: this);
          case 'custom_scroll_view':
            return _CustomScrollView(
                data: data, parentData: parentData, componentFactory: this);
          case 'editable_text':
            return _EditableText(
                data: data, parentData: parentData, componentFactory: this);
          case 'decorated_box':
            return _DecoratedBox(
                data: data,
                parentData: parentData,
                componentFactory: this,
                isFront: false);
          case 'foreground_decorated_box':
            return _DecoratedBox(
                data: data,
                parentData: parentData,
                componentFactory: this,
                isFront: true);
          case 'sliver_list':
            return _SliverList(
                data: data, parentData: parentData, componentFactory: this);
          case 'sliver_grid':
            return _SliverGrid(
                data: data, parentData: parentData, componentFactory: this);
          case 'sliver_persistent_header':
            return _SliverPersistentHeader(
                data: data, parentData: parentData, componentFactory: this);
          case 'gesture_detector':
            return _GestureDetector(
                data: data, parentData: parentData, componentFactory: this);
          case 'grid_view':
            return _GridView(
                data: data, parentData: parentData, componentFactory: this);
          case 'ignore_pointer':
            return _IgnorePointer(
                data: data, parentData: parentData, componentFactory: this);
          case 'image':
            return _Image(
                data: data, parentData: parentData, componentFactory: this);
          case 'list_view':
            return _ListView(
                data: data, parentData: parentData, componentFactory: this);
          case 'opacity':
            return _Opacity(
                data: data, parentData: parentData, componentFactory: this);
          case 'overlay':
            return _Overlay(
                data: data, parentData: parentData, componentFactory: this);
          case 'rich_text':
            return _RichText(
                data: data, parentData: parentData, componentFactory: this);
          case 'offstage':
            return _Offstage(
                data: data, parentData: parentData, componentFactory: this);
          case 'transform':
            return _Transform(
                data: data, parentData: parentData, componentFactory: this);
          case 'visibility':
            return _Visibility(
                data: data, parentData: parentData, componentFactory: this);
          case 'mp_scaffold':
            return _MPScaffold(
                data: data, parentData: parentData, componentFactory: this);
          case 'mp_date_picker':
            return _MPDatePicker(
                data: data, parentData: parentData, componentFactory: this);
          case 'mp_picker':
            return _MPPicker(
                data: data, parentData: parentData, componentFactory: this);
          case 'mp_icon':
            return _MPIcon(
                data: data, parentData: parentData, componentFactory: this);
          case 'mp_page_view':
            return _MPPageView(
                data: data, parentData: parentData, componentFactory: this);
          case 'mp_switch':
            return _MPSwitch(
                data: data, parentData: parentData, componentFactory: this);
          case 'mp_slider':
            return _MPSlider(
                data: data, parentData: parentData, componentFactory: this);
          case 'mp_circular_progress_indicator':
            return _MPCircularProgressIndicator(
                data: data, parentData: parentData, componentFactory: this);
          default:
            if (MPPluginRegister.registedViews.containsKey(name)) {
              return MPPluginRegister.registedViews[name]!(
                null,
                data,
                parentData,
                this,
              );
            }
            return ComponentView(
                data: data, parentData: parentData, componentFactory: this);
        }
      }
    }
    return Container();
  }

  clear() {
    _textMeasureResults.clear();
    _cacheViews.clear();
  }

  void _callbackTextMeasureResult(dynamic measureId, Size size) {
    _textMeasureResults.add({
      'measureId': measureId,
      'size': {'width': size.width, 'height': size.height},
    });
  }

  void _callbackTextPainterMeasureResult(dynamic measureId, Size size) {
    engine._sendMessage({
      'type': 'rich_text',
      'message': {
        'event': 'onTextPainterMeasured',
        'data': {
          'seqId': measureId,
          'size': {'width': size.width, 'height': size.height},
        },
      },
    });
  }

  void _flushTextMeasureResult() {
    if (_textMeasureResults.isNotEmpty) {
      engine._sendMessage({
        'type': 'rich_text',
        'message': {
          'event': 'onMeasured',
          'data': _textMeasureResults,
        },
      });
      _textMeasureResults.clear();
    }
  }
}

File: packages/mp_dom_runtime/src/components/component_factory.ts
import { Engine } from "../engine";
import { AbsorbPointer } from "./basic/absorb_pointer";
import { ClipOval } from "./basic/clip_oval";
import { ClipRRect, ClipRRectAncestor } from "./basic/clip_r_rect";
import { ColoredBox } from "./basic/colored_box";
import { DecoratedBox } from "./basic/decorated_box";
import { GestureDetector } from "./basic/gesture_detector";
import { IgnorePointer } from "./basic/ignore_pointer";
import { Offstage } from "./basic/offstage";
import { Opacity, OpacityAncestor } from "./basic/opacity";
import { Transform } from "./basic/transform";
import { Image } from "./basic/image";
import { AncestorView, ComponentView } from "./component_view";
import { MPScaffold } from "./mpkit/scaffold";
import { RichText, TextSpan, WidgetSpan } from "./basic/rich_text";
import { ListView } from "./basic/list_view";
import { GridView } from "./basic/grid_view";
import { CustomScrollView } from "./basic/custom_scroll_view";
import { CustomPaint } from "./basic/custom_paint";
import { MPIcon } from "./mpkit/icon";
import { MPVideoView } from "./mpkit/video_view";
import { MPWebView } from "./mpkit/web_view";
import { MPPageView } from "./mpkit/page_view";
import { MPMiniProgramView } from "./mpkit/miniprogram_view";
import { Overlay } from "./basic/overlay";
import { Visibility } from "./basic/visibility";
import { SliverPersistentHeader } from "./basic/sliver_persistent_header";
import { EditableText } from "./basic/editable_text";
import { MPPlatformView } from "./mpkit/platform_view";
import { ForegroundDecoratedBox } from "./basic/foreground_decorated_box";
import { MPEnv } from "../env";
import { MPSwitch } from "./mpkit/switch";
import { MPSlider } from "./mpkit/slider";
import { MPPicker } from "./mpkit/picker";
import { MPDatePicker } from "./mpkit/date_picker";
import { MPCircularProgressIndicator } from "./mpkit/circular_progress_indicator";
import { MouseRegion } from "./basic/mouse_region";

export class ComponentFactory {
  static components: { [key: string]: typeof ComponentView } = {
    absorb_pointer: AbsorbPointer,
    clip_oval: ClipOval,
    clip_r_rect: ClipRRect,
    colored_box: ColoredBox,
    custom_paint: CustomPaint,
    custom_scroll_view: CustomScrollView,
    decorated_box: DecoratedBox,
    foreground_decorated_box: ForegroundDecoratedBox,
    editable_text: EditableText,
    gesture_detector: GestureDetector,
    grid_view: GridView,
    ignore_pointer: IgnorePointer,
    image: Image,
    list_view: ListView,
    mouse_region: MouseRegion,
    offstage: Offstage,
    opacity: Opacity,
    overlay: Overlay,
    rich_text: RichText,
    text_span: TextSpan,
    widget_span: WidgetSpan,
    sliver_persistent_header: SliverPersistentHeader,
    transform: Transform,
    visibility: Visibility,
    mp_icon: MPIcon,
    mp_scaffold: MPScaffold,
    mp_video_view: MPVideoView,
    mp_web_view: MPWebView,
    mp_page_view: MPPageView,
    mp_platform_view: MPPlatformView,
    mp_switch: MPSwitch,
    mp_slider: MPSlider,
    mp_picker: MPPicker,
    mp_date_picker: MPDatePicker,
    mp_mini_program_view: MPMiniProgramView,
    mp_circular_progress_indicator: MPCircularProgressIndicator,
  };

  static ancestors: { [key: string]: typeof AncestorView } = {
    opacity: OpacityAncestor,
    clip_r_rect: ClipRRectAncestor,
  };

  static disableCache = false;

  cachedView: { [key: number]: ComponentView } = {};
  cachedAncestor: { [key: number]: AncestorView } = {};
  cachedElement: { [key: number]: any } = {};
  private textMeasureResults: {
    measureId: number;
    size: { width: number; height: number };
  }[] = [];

  constructor(readonly engine: Engine) {}

  create(data: any, document: Document): ComponentView | undefined {
    if (!data) return undefined;
    const same = data["^"];
    const name = data.name;
    const hashCode = data.hashCode;
    if (same == 1 && typeof hashCode === "number") {
      return this.cachedView[hashCode];
    }
    if (!name || !hashCode) {
      return undefined;
    }
    if (!same) {
      this.cachedElement[hashCode] = data;
    }
    const cachedView = !ComponentFactory.disableCache && this.cachedView[hashCode];
    if (cachedView) {
      document = cachedView.document;
      if (data.ancestors) {
        cachedView.setAncestors(data.ancestors);
      }
      if (data.constraints) {
        cachedView.setConstraints(data.constraints);
      }
      if (data.attributes) {
        cachedView.setAttributes(data.attributes);
      }
      if (data.children) {
        cachedView.setChildren(this.fetchCachedChildren(data.children));
      }
      return cachedView;
    }
    if (!document) return;
    let clazz = ComponentFactory.components[name];
    if (!clazz) {
      clazz = ComponentView;
    }
    const view = new clazz(document, data.attributes);
    if (MPEnv.platformGlobal()?.mpDEBUG) {
      view.htmlElement.setAttribute("mp-component", name);
    }
    view.factory = this;
    view.engine = this.engine;
    view.hashCode = hashCode;
    if (data.ancestors) {
      view.setAncestors(data.ancestors);
    }
    if (data.constraints) {
      view.setConstraints(data.constraints);
    }
    if (data.attributes) {
      view.setAttributes(data.attributes);
    }
    if (data.children) {
      if (ComponentFactory.disableCache) {
        view.setChildren(data.children);
      } else {
        view.setChildren(this.fetchCachedChildren(data.children));
      }
    }
    if (!ComponentFactory.disableCache) {
      this.cachedView[hashCode] = view;
    }
    return view;
  }

  createAncestors(data: any, target: ComponentView): AncestorView | undefined {
    const same = data["^"];
    const name = data.name;
    const hashCode = data.hashCode;
    if (same == 1 && typeof hashCode === "number") {
      return this.cachedAncestor[hashCode];
    }
    if (!name || !hashCode) {
      return undefined;
    }
    if (!same) {
      this.cachedElement[hashCode] = data;
    }
    const cachedAncestor = this.cachedAncestor[hashCode];
    if (cachedAncestor) {
      if (cachedAncestor.target && cachedAncestor.target !== target) {
        const idx = cachedAncestor.target.ancestors.indexOf(cachedAncestor);
        if (idx >= 0) cachedAncestor.target.ancestors.splice(idx, 1);
      }
      cachedAncestor.target = target;
      if (data.attributes) {
        cachedAncestor.setAttributes(data.attributes);
      }
      if (data.constraints) {
        cachedAncestor.setConstraints(data.constraints);
      }
      return cachedAncestor;
    }
    if (!target) return;
    let clazz = ComponentFactory.ancestors[name];
    if (!clazz) {
      return undefined;
    }
    const ancestor = new clazz(target);
    if (data.attributes) {
      ancestor.setAttributes(data.attributes);
    }
    if (data.constraints) {
      ancestor.setConstraints(data.constraints);
    }
    this.cachedAncestor[hashCode] = ancestor;
    return ancestor;
  }

  fetchCachedChildren(children: any[]) {
    return children.map((it: any) => {
      let same = it["^"];
      let hashCode = it["hashCode"];
      if (same && this.cachedElement[hashCode]) {
        return this.cachedElement[hashCode];
      } else {
        return it;
      }
    });
  }

  private markedNeedsFlushTextMeasureResult = false;

  callbackTextMeasureResult(measureId: number, size: { width: number; height: number }) {
    this.textMeasureResults.push({ measureId, size });
    if (!this.markedNeedsFlushTextMeasureResult) {
      this.markedNeedsFlushTextMeasureResult = true;
      setTimeout(() => {
        this.flushTextMeasureResult();
        this.markedNeedsFlushTextMeasureResult = false;
      }, 50);
    }
  }

  callbackTextPainterMeasureResult(seqId: number, size: { width: number; height: number }) {
    this.engine.sendMessage(
      JSON.stringify({
        type: "rich_text",
        message: { event: "onTextPainterMeasured", data: { seqId, size } },
      })
    );
  }

  flushTextMeasureResult() {
    if (this.textMeasureResults.length > 0) {
      this.engine.sendMessage(
        JSON.stringify({
          type: "rich_text",
          message: { event: "onMeasured", data: this.textMeasureResults },
        })
      );
      this.textMeasureResults = [];
    }
  }
}
